### PR TITLE
[WIP] Support gym.spaces.Dict

### PR DESCRIPTION
**Pull request by takerfume**

### CODE REVIEW

**Comment by takerfume** [comment]
All that is left is to support a nested dictionary of gym.spaces.Dict like below.
```
    Example usage [nested]:
    self.nested_observation_space = spaces.Dict({
        'sensors':  spaces.Dict({
            'position': spaces.Box(low=-100, high=100, shape=(3)),
            'velocity': spaces.Box(low=-1, high=1, shape=(3)),
            'front_cam': spaces.Tuple((
                spaces.Box(low=0, high=1, shape=(10, 10, 3)),
                spaces.Box(low=0, high=1, shape=(10, 10, 3))
            )),
            'rear_cam': spaces.Box(low=0, high=1, shape=(10, 10, 3)),
        }),
        'ext_controller': spaces.MultiDiscrete([ [0,4], [0,1], [0,1] ]),
        'inner_state':spaces.Dict({
            'charge': spaces.Discrete(100),
            'system_checks': spaces.MultiBinary(10),
            'job_status': spaces.Dict({
                'task': spaces.Discrete(5),
                'progress': spaces.Box(low=0, high=100, shape=()),
            })
        })
    })
```

**Comment by takerfume** [comment]
I will leave supporting nested gym.spaces.Dict described my above comment for another PR.
I have finished writing code in this PR.
I have finished writing the following contents.
* Use `gym.wrappers.FlattenDictWrapper` in order to transform gym.spaces.Dict to flatten gym.spaces.Box
* Define the function of `machina.envs.flatten_to_dict` in order to transform flatten obs to dict obs.
* Test the consistency of value of obs in `tests.test_env.test_flatten2dict`
* Test if dict obs can be used to train ppo and sac in `tests.test_env.TestFlatten2DictPP0` and `tests.test_env.TestFlatten2DictSAC`

**Comment by takerfume** [comment]
@rarilurelo @iory Could you check this PR?

File: machina/envs/env_utils.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    begin_index = 0
+    end_index = 0
+    for key in dict_keys:
+        origin_shape = dict_space.spaces[key].shape
+        end_index += np.prod(origin_shape)
+        dim = len(flatten_obs.shape)
+        if dim == 2:  # not rnn
+            obs_dict[key] = flatten_obs[:, begin_index:end_index].reshape(
+                flatten_obs.shape[:-1] + origin_shape)
+        elif dim == 3:  # rnn
```
Comment (rarilurelo) [inline machina/envs/env_utils.py]: Is this if statement can support rnn with sampling for example R2D2-SAC? (timeseq, sampling, batch_size, ob_dim)
I think this statement should be written as `flatten_obs[..., begin_index:end_index]`.

Comment (takerfume) [reply machina/envs/env_utils.py]: yes.
I tested rnn.
I will try it based on your idea.

**Comment by takerfume** [comment]
@rarilurelo I added test of r2d2.

**Pull request merged**